Packs/OneLogin/Integrations/OneLoginEventCollector/OneLoginEventCollector.py

```diff
@@ -0,0 +1,367 @@
+                    events = events[idx:]
+                    break
+            last_run.pop('first_id', None)  # removing to make sure it won't be used in future runs
+
+        return events, cursor
+
+    def fetch_events(self, access_token: str, query_params: dict, last_run: dict) -> List[dict]:
+        """
+        Aggregates events using cursor-based pagination, until one of the following occurs:
+        1. Reaches the user-defined limit (parameter).
+           In this case, stores the last used `cursor` and `since` and the id of the next event to collect (`first_id`)
+           and returns the events that have been accumulated so far.
+
+        2. Reaches the end of the pagination.
+           In this case, the lastRun obj will be updated with the `cursor` that will be None
+           and with a new value for `since` taken from the last event that were fetched.
+
+        3. Reaches a rate limit.
+           In this case, stores the last cursor used in the lastRun obj
+           and returns the events that have been accumulated so far.
+        """
+        aggregated_events: List[dict] = []
+
+        event_types = self.get_event_types_from_last_run(last_run)
+
+        user_defined_limit = query_params.pop('limit')
+        query_params['limit'] = 1000  # Constant limit value for the request
+        try:
+            events, cursor = self.handle_pagination_first_batch(access_token, query_params, last_run)
+            while events:
+                for event in events:
+
+                    if len(aggregated_events) == user_defined_limit:
+                        demisto.debug(f'Reached the user-defined limit ({user_defined_limit}) - stopping.')
+                        last_run['first_id'] = event.get('id')
+                        cursor = query_params['after_cursor']
+                        break
+
+                    event['event_type_name'] = self.convert_type_id_to_name(event, event_types, last_run)
+                    aggregated_events.append(event)
+
+                else:
+                    # Finished iterating through all events in this batch
+                    if cursor:
+                        demisto.debug('Using the cursor from the last API call to execute the next call.')
+                        query_params['after_cursor'] = cursor
+                        _, events, cursor = self.get_events_request(access_token, query_params)
+                        continue
+
+                demisto.debug('Finished iterating through all events in this fetch run.')
+                break
+
+        except DemistoException as e:
+            if not e.res or e.res.status_code != 429:
+                raise e
+            demisto.debug('Reached API rate limit, storing last used cursor.')
+            cursor = query_params['after_cursor']
+
+        since = query_params['since'] if cursor or not aggregated_events else increase_time_with_second(
+            aggregated_events[-1].get('created_at'))
```

Comment:
is it possible to miss events like this?

Reply:
is it possible to have multiple events at the exact same time in the product? then if we fetch only some of them in the first run, we won't fetch the rest of them in the following run